lib/iati_validator/xml.rb

```diff
@@ -0,0 +1,62 @@
+require "nokogiri"
+
+module IATIValidator
+  class XML
+    class InvalidError < StandardError
+      attr_reader :errors, :url
+      def initialize(errors, url)
+        @errors = errors
+        @url = url
+      end
+
+      def skip_validation_url
+        url = Addressable::URI.parse(@url)
+        query = url.query_values || {}
+        url.query_values = query.merge(skip_validation: 1)
+        url
+      end
+
+      def to_s
+        "XML validation errors were found in #{skip_validation_url}\n#{@errors.join("\n")}"
+      end
+    end
+
+    SCHEMAS = {"iati-activities" => "iati-activities",
+               "iati-organisations" => "iati-organisations",
+               "registry-record" => "iati-registry-record"}
+    PATH = "vendor/data/IATI-Schemas-2.03/%s-schema.xsd"
+
+    attr_reader(:errors)
+
+    def initialize(content)
+      @content = content
+      @errors = []
+    end
+
+    def error(message)
+      @errors << message
+      false
+    end
+
+    def valid?
+      doc = Nokogiri::XML(@content)
+
+      return error("Not XML. #{doc.errors}") unless doc.errors.empty?
+      return error("No root node.") unless doc.root
+
+      schema = SCHEMAS[doc.root.name]
+      return error("Did not recognise opening tag #{doc.root.name}") unless schema
+
+      xsd = Nokogiri::XML::Schema(File.open(PATH % schema))
+
+      validation_errors = xsd.validate(doc)
+      unless validation_errors.empty?
```

Comment:
Prefer the positive case with the antonym of `#empty?`: `if validation_errors.any?`

Reply:
(or could go with `return true if validation_errors.empty?`)

Reply:
<img width="220" alt="image" src="https://user-images.githubusercontent.com/194511/160642754-11fdc085-4e9f-443b-9259-a30bcd03c8c9.png">
^^ this indentation, for the avoidance of doubt :wink:

Reply:
:+1: